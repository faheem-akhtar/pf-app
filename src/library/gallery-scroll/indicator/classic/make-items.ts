import { arrayFromRange } from 'helpers/array/from-range';

import { GalleryScrollIndicatorItemInterface } from '../item.interface';
import { GALLERY_SCROLL_INDICATOR_CLASSIC_VISIBLE_ITEMS_COUNT } from './visible-items-count.constant';

export const galleryScrollIndicatorClassicMakeItems = (
  itemsCount: number,
  activeIndex: number
): GalleryScrollIndicatorItemInterface[] =>
  arrayFromRange(0, itemsCount).map((i) => {
    const isActive = i === activeIndex;
    let isVisible;
    const firstActive = activeIndex === 0;
    const lastActive = activeIndex === itemsCount - 1;
    const numberOfSmallDots = GALLERY_SCROLL_INDICATOR_CLASSIC_VISIBLE_ITEMS_COUNT - 1;
    if (firstActive) {
      isVisible = i <= numberOfSmallDots;
    } else if (lastActive) {
      isVisible = i >= activeIndex - numberOfSmallDots;
    } else {
      isVisible = Math.abs(i - activeIndex) <= numberOfSmallDots / 2;
    }

    const isHidden = !isVisible;

    return {
      isActive,
      isHidden,
      isSmall: !isHidden && !isActive,
    };
  });
