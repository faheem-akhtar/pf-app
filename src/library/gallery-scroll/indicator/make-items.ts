import { arrayFromRange } from 'helpers/array/from-range';
import { GalleryScrollIndicatorItemInterface } from './item.interface';
import { galleryScrollIndicatorVisibleItemsCount } from './visible-items-count';

export const galleryScrollIndicatorMakeItems = (
  itemsCount: number,
  activeIndex: number
): GalleryScrollIndicatorItemInterface[] =>
  arrayFromRange(0, itemsCount).map((i) => {
    const isActive = i === activeIndex;
    let isVisible;
    const firstActive = activeIndex === 0;
    const lastActive = activeIndex === itemsCount - 1;
    const numberOfSmallDots = galleryScrollIndicatorVisibleItemsCount - 1;
    if (firstActive) {
      isVisible = i <= numberOfSmallDots;
    } else if (lastActive) {
      isVisible = i >= activeIndex - numberOfSmallDots;
    } else {
      isVisible = Math.abs(i - activeIndex) <= numberOfSmallDots / 2;
    }

    const isHidden = !isVisible;

    return {
      isActive,
      isHidden,
      isSmall: !isHidden && !isActive,
    };
  });
