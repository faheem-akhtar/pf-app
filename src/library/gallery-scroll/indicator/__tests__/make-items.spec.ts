import { galleryScrollIndicatorMakeItems } from '../make-items';

describe('galleryScrollIndicatorMakeItems', () => {
  it('3 items, last is active', () => {
    const result = galleryScrollIndicatorMakeItems(3, 2);

    expect(result).toMatchSnapshot();
  });

  it('3 items, first is active', () => {
    const result = galleryScrollIndicatorMakeItems(3, 0);

    expect(result).toMatchSnapshot();
  });

  it('3 items, second is active', () => {
    const result = galleryScrollIndicatorMakeItems(3, 1);

    expect(result).toMatchSnapshot();
  });
});
