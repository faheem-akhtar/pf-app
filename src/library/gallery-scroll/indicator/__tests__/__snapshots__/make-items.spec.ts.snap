// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`galleryScrollIndicatorMakeItems 3 items, first is active 1`] = `
Array [
  Object {
    "isActive": true,
    "isHidden": false,
    "isSmall": false,
  },
  Object {
    "isActive": false,
    "isHidden": false,
    "isSmall": true,
  },
  Object {
    "isActive": false,
    "isHidden": false,
    "isSmall": true,
  },
]
`;

exports[`galleryScrollIndicatorMakeItems 3 items, last is active 1`] = `
Array [
  Object {
    "isActive": false,
    "isHidden": false,
    "isSmall": true,
  },
  Object {
    "isActive": false,
    "isHidden": false,
    "isSmall": true,
  },
  Object {
    "isActive": true,
    "isHidden": false,
    "isSmall": false,
  },
]
`;

exports[`galleryScrollIndicatorMakeItems 3 items, second is active 1`] = `
Array [
  Object {
    "isActive": false,
    "isHidden": false,
    "isSmall": true,
  },
  Object {
    "isActive": true,
    "isHidden": false,
    "isSmall": false,
  },
  Object {
    "isActive": false,
    "isHidden": false,
    "isSmall": true,
  },
]
`;
